packages/webpack-cli/src/plugins/dotenv-webpack-plugin.ts

```diff
@@ -0,0 +1,196 @@
+  matches.forEach((match) => {
+    env = env.replace(match, interpolate(vars[match.replace(/\$|{|}/g, "")] || "", vars));
+  });
+  return env;
+};
+
+const isMainThreadElectron = (target: string | undefined): boolean =>
+  !!target && target.startsWith("electron") && target.endsWith("main");
+
+export class Dotenv {
+  #logger!: Logger;
+  #config: DotenvConfig;
+  #inputFileSystem: any;
+  #cached: boolean;
+
+  constructor(config: DotenvConfig = {}) {
+    const nodeEnv = process.env.NODE_ENV || "development";
+
+    this.#config = {
+      paths: [
+        ".env",
+        ...(nodeEnv === "test" ? [] : [".env.local"]),
+        `.env.${nodeEnv}`,
+        `.env.${nodeEnv}.local`,
+      ],
+      prefixes: ["process.env.", "import.meta.env."],
+      ...config,
+    };
+    this.#cached = false;
+  }
+
+  public apply(compiler: Compiler): void {
+    this.#inputFileSystem = compiler.inputFileSystem;
+    this.#logger = compiler.getInfrastructureLogger("dotenv-webpack-plugin");
+    compiler.hooks.thisCompilation.tap("dotenv-webpack-plugin", async (compilation) => {
+      const cache = compilation.getCache("dotenv-webpack-plugin-compiler");
+
+      if (!this.#cached) {
+        const variables = this.#gatherVariables(compilation);
+        const target: string =
+          typeof compiler.options.target == "string" ? compiler.options.target : "";
+        const data = this.#formatData({
+          variables,
+          target: target,
+        });
+        new DefinePlugin(data).apply(compiler);
+        await cache
+          .storePromise("dotenv-webpack-plugin-data", null, data)
+          .then(() => (this.#cached = true));
+      } else {
+        new DefinePlugin(await cache.getPromise("dotenv-webpack-plugin-data", null)).apply(
+          compiler,
+        );
+      }
+    });
+  }
+  #gatherVariables(compilation: any): EnvVariables {
+    const { allowEmptyValues } = this.#config;
+    const vars: EnvVariables = this.#initializeVars();
+
+    const { env, blueprint } = this.#getEnvs(compilation);
+
+    Object.keys(blueprint).forEach((key) => {
+      const value = Object.prototype.hasOwnProperty.call(vars, key) ? vars[key] : env[key];
+
+      if (typeof value === "undefined" || value === null || (!allowEmptyValues && value === "")) {
+        compilation.errors.push(new Error(`Missing environment variable: ${key}`));
+      } else {
+        vars[key] = value;
+      }
+    });
+
+    return vars;
+  }
+
+  #initializeVars(): EnvVariables {
+    return this.#config.systemvars
+      ? Object.fromEntries(Object.entries(process.env).map(([key, value]) => [key, value ?? ""]))
+      : {};
+  }
+
+  #getEnvs(compilation: any): { env: EnvVariables; blueprint: EnvVariables } {
+    const { paths } = this.#config;
+
+    const env: EnvVariables = {};
+
+    (paths || ["./.env"]).forEach((path) =>
+      Object.assign(env, dotenv.parse(this.#loadFile(path || "./.env", compilation))),
```

Comment:
Avoid fallback to `"./.env"`, just throw an error if there are no paths in `this.#config`